js/libs/form.js

```diff
@@ -472,185 +474,241 @@ Fliplet.Widget.instance('form-builder', function(data) {
 
             if (inputField.$v.$invalid) {
               $(inputField.$el).addClass('has-error');
+              invalidFields.push(inputField);
               $vm.isFormValid = false;
             }
           }
         });
 
-       if(!$vm.isFormValid){
-         return;
-       }
-
-        this.isSending = true;
+        /**
+         * Showing an error message
+         * 
+         * @param {String} errorMessage - an error message that we should show to the user
+         *  if its empty string show default message
+         * @returns {void} shows a toast message to users
+         */
+        function showValidationMessage(errorMessage) {
+          errorMessage = errorMessage || 'Please complete all required fields.';
+          Fliplet.UI.Toast(errorMessage);
+        }
 
-        function appendField(name, value) {
-          if (Array.isArray(formData[name])) {
-            formData[name].push(value);
-          } else if (typeof formData[name] !== 'undefined') {
-            formData[name] = [formData[name], value];
-          } else {
-            formData[name] = value;
-          }
+        /**
+         * This method will decide what we will do after isFormInvalid hook
+         * 
+         * @returns {Promise} With this logic:
+         *  1. In case there was no listener on isFormInvalid hook we will show the toast
+         *     with default message
+         *  2. In case when listener was resolved with Promise.resolve() we allow user to submit invalid form
+         *     and will not show the toast
+         *  3. In case when listener was resolved with Promise.reject() we will not show the toast
+         *     and not allow form to submit
+         *  4. In case when listener was resolved with Promise.reject('') we will show the toast with default text
+         *  5. In case when listener was resolved with Promise.reject('error text') we will show the toast
+         *     with 'error text' message
+         */
+        function onFormInvalid() {
+          return new Promise(function(resolve, reject) {
+            Fliplet.Hooks.run('isFormInvalid', invalidFields)
```

Comment:
@tonytlwu @YaroslavOvdii I don't think the wrapping promise here was necessary: `run` already returns a promise.

Reply:
You're right.